consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,436 @@
+	// TotalVotingPower represents N which is required to calculate the thresholds.
+	TotalVotingPower(height uint) uint
+
+	// ValidatorVotingPower returns the voting power of the a single validator. This is also required to implement
+	// various thresholds. The assumption is that a single validator cannot have voting power more than f.
+	ValidatorVotingPower(validatorAddr A) uint
+
+	// Proposer returns the proposer of the current round and height.
+	Proposer(height, round uint) A
+}
+
+type Slasher[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
+	// Equivocation informs the slasher that a validator has sent conflicting messages. Thus it can decide whether to
+	// slash the validator and by how much.
+	Equivocation(msgs ...M)
+}
+
+type Listener[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
+	// Listen would return consensus messages to Tendermint which are set // by the validator set.
+	Listen() <-chan M
+}
+
+type Broadcaster[M Message[V, H, A], V Hashable[H], H Hash, A Addr] interface {
+	// Broadcast will broadcast the message to the whole validator set. The function should not be blocking.
+	Broadcast(msg M)
+
+	// SendMsg would send a message to a specific validator. This would be required for helping send resquest and
+	// response message to help a specifc validator to catch up.
+	SendMsg(validatorAddr A, msg M)
+}
+
+type Listeners[V Hashable[H], H Hash, A Addr] struct {
+	ProposalListener  Listener[Proposal[V, H, A], V, H, A]
+	PrevoteListener   Listener[Prevote[H, A], V, H, A]
+	PrecommitListener Listener[Precommit[H, A], V, H, A]
+}
+
+type Broadcasters[V Hashable[H], H Hash, A Addr] struct {
+	ProposalBroadcaster  Broadcaster[Proposal[V, H, A], V, H, A]
+	PrevoteBroadcaster   Broadcaster[Prevote[H, A], V, H, A]
+	PrecommitBroadcaster Broadcaster[Precommit[H, A], V, H, A]
+}
+
+type Tendermint[V Hashable[H], H Hash, A Addr] struct {
+	nodeAddr A
+
+	state state[V, H] // Todo: Does state need to be protected?
+
+	messages       messages[V, H, A]
+	futureMessages messages[V, H, A]
+
+	futureMessagesMu *sync.Mutex
+
+	timeoutPropose   timeoutFn
+	timeoutPrevote   timeoutFn
+	timeoutPrecommit timeoutFn
+
+	application Application[V, H]
+	blockchain  Blockchain[V, H, A]
+	validators  Validators[A]
+
+	listeners    Listeners[V, H, A]
+	broadcasters Broadcasters[V, H, A]
+
+	scheduledTms []timeout
+	timeoutsCh   chan timeout
+
+	// Future round messages are sent to the loop through the following channels
+	proposalsCh  chan Proposal[V, H, A]
+	prevotesCh   chan Prevote[H, A]
+	precommitsCh chan Precommit[H, A]
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+type state[V Hashable[H], H Hash] struct {
+	height uint
+	round  uint
+	step   step
+
+	lockedValue *V
+	lockedRound int
+	validValue  *V
+	validRound  int
+
+	// The following are round level variable therefore when a round changes they must be reset.
+	line34Executed bool
+	line36Executed bool
+	line47Executed bool
```

Comment:
I understand that we're strictly following the pseudocode, but I feel that this is not easy to read and memorize. Is it possible to name these differently?